### ISSUE TITLE
Support for cpu_shares

### ISSUE THREAD

**Issue by sstarcher** [kubernetes-incubator/kompose#267]
It would be nice to have resources limits for cpu_shares and mem_limits.

**Comment by containscafeine** [comment]
It would be great to have support for [mem_limit](https://docs.docker.com/compose/compose-file/#/cpushares-cpuquota-cpuset-domainname-hostname-ipc-macaddress-memlimit-memswaplimit-oomscoreadj-privileged-readonly-restart-shmsize-stdinopen-tty-user-workingdir) which should translate to a pod's spec like -

```yaml
spec:
  containers:
  - image: nginx
    imagePullPolicy: Always
    name: nginx
    resources:
      limits:
        memory: 200Mi
    terminationMessagePath: /dev/termination-log
    volumeMounts:
```

**Comment by surajssd** [comment]
@containscafeine :+1:

**Comment by cdrage** [comment]
I'll start working on this. I'll assign myself to this issue!

**Comment by cdrage** [comment]
Assigning to @containscafeine instead.

**Comment by containscafeine** [comment]
Sigh, okay, so taking a look at the `mem_limit` part, and having implemented it, turns out that if the docker-compose.yaml looks like -
```yaml
web:
  image: tuna/docker-counter23
  ports:
    - "5000:5000"
  links:
    - redis
  mem_limit: 500m
```

i.e., has a non integer value, then, `kompose convert` fails with -

```shell
ERRO[0000] Could not parse config for project tmp : strconv.ParseInt: parsing "500m": invalid syntax 
FATA[0000] Failed to load compose file: strconv.ParseInt: parsing "500m": invalid syntax 
```

Tracking this down further, this is a libcompose issue and `libcompose up` fails with -

```shell
ERRO[0000] Failed to unmarshall: strconv.ParseInt: parsing "500m": invalid syntax
web:
  image: tuna/docker-counter23
  links:
  - redis
  mem_limit: 500m
  ports:
  - 5000:5000 
ERRO[0000] Could not parse config for project tmp : strconv.ParseInt: parsing "500m": invalid syntax 
FATA[0000] Failed to read project: strconv.ParseInt: parsing "500m": invalid syntax 
```

However, trying `libcompose up` with the [latest release](https://github.com/docker/libcompose/releases/tag/v0.4.0) works fine.

So, for this to work fine, libcompose version has to be bumped.

**Comment by kadel** [comment]
@containscafeine can you test it with  https://github.com/kubernetes-incubator/kompose/pull/356?

**Comment by containscafeine** [comment]
@cdrage just tried this with #356, and mem_limit fails. This is fixed in the latest release of [`libcompose`](https://github.com/docker/libcompose/releases/tag/v0.4.0), can we upgrade to that?

**Comment by surajnarwade** [comment]
@cdrage @surajssd we can close this issue, as mem_limit is now supported

**Comment by cdrage** [comment]
@surajnarwade cpu_shares is still an issue. So let's leave this open.

**Comment by surajnarwade** [comment]
@cdrage oh my bad, got it

**Comment by surajssd** [comment]
@cdrage `mem_limit` is supported right now so edited the title

**Comment by surajssd** [comment]
@cdrage @surajnarwade we can officially declare that we don't support `cpu_shares` becasue there is no way we can map it to anything in k8s.

`cpu_shares` talks in terms of percentage of cpu to give to any container, this is feasible on a single machine but does not make sense on multi-node cluster where we dont know the entire cpu number.

TLDR: this is not easy to map to k8s cpu resources.

more info about cpu shares: https://docs.docker.com/engine/admin/resource_constraints/#cpu

**Comment by surajssd** [comment]
So the PR that closes this issue should remove all the code which tries to read this info from docker compose file and adds it to `ServiceConfig` and also give warning as this is unsupported.

**Comment by cdrage** [comment]
@surajssd Adding this to unsupportedkeys sounds like the best viable option considering the complexity of implementing cpu_shares. I agree.

**Comment by surajnarwade** [comment]
@surajssd @cdrage , I am onto this.

**Comment by surajnarwade** [comment]
we can close this now

**Comment by gitlawr** [comment]
Hope not to be pedantic here, but I think for `cpu_shares` and `cpu_quota` there are highly relevant concepts in Kubernetes: [see](https://kubernetes.io/docs/concepts/configuration/manage-compute-resources-container/#how-pods-with-resource-limits-are-run).
These two keys in compose file are analogous to `docker run` counterpart, while in k8s `requests.cpu` maps to `docker run --cpu-shares` and `limits.cpu` maps to `--cpu-quota`(involving simple computations and pre-condition that `--cpu-period=100000` though)
A user case I can come up with is that we may use daemonset or pod affinity to run one instance on every node and keep dedicated share/quota of CPU for it.
As for `cpuset` key I agree that it is pointless in a cluster environment as we usually don't care which cpu is used.
WDYT? @surajssd @cdrage @surajnarwade

**Comment by surajnarwade** [comment]
@gitlawr , but there is no direct mapping for `cpu_shares` and `quotas`

**Comment by gitlawr** [comment]
@surajnarwade For example,setting cpu resource such as:
```
apiVersion: v1
kind: Pod
metadata:
  name: one
spec:
  volumes:
  containers:
  - name: busysleep
    image: busybox
    command: ["sleep","3000"]
    resources:
      requests:
        cpu: 100m
      limits:
        cpu: 0.5
```
Then the container will be run with CpuShares==102(which is 0.1*1024) and CpuQuota=50000(which is half of CpuPeriod given CpuPeriod= 100000). Is it kind of direct mapping I think?
